### ISSUE TITLE
cxl create-region : Failed to synchronize CPU cache state (QEMU)

### ISSUE THREAD

**Issue by shadowcollecter** [pmem/ndctl#250]
When I want to create a region in CXL Type3 device , I faced this issue.

```
# cxl create-region -d decoder0.0 -t pmem -m mem0
[   72.310604] cxl_region region0: Failed to synchronize CPU cache state
libcxl: cxl_region_enable: region0: failed to enable
cxl region: create_region: region0: failed to enable: No such device or address
cxl region: cmd_create_region: created 0 regions
```

It is interesting that when I switch to kernel 6.5 (kernel-6.5.0-0.rc1.11.fc39)
the terminal shows the same error and freeze(not happen in the 6.3 kernel).

CXL-CLI version  : 77
QEMU version  : 8.0.50
Host CPU : Xeon Cascade Lake (Google Cloud instance :  n2-standard-8) 
QEMU VM kernel version : kernel-6.3.13-200.fc38

```
QEMU Command (as similar as the official doc)
$ sudo ~/qemu-v4/build/qemu-system-x86_64  \ 
-drive file=~/disk/Fedora-Server-KVM-38-1.6.x86_64.qcow2,format=qcow2,index=0,media=disk,id=hd \
-m 4G,slots=8,maxmem=8G -smp 4 -machine type=q35,cxl=on  \
-nographic -enable-kvm \
-object memory-backend-ram,id=pmem0,share=on,size=256M \
-object memory-backend-file,id=cxl-lsa0,share=on,mem-path=/tmp/lsa.raw,size=256M \
-device pxb-cxl,bus_nr=12,bus=pcie.0,id=cxl.1 \
-device cxl-rp,port=0,bus=cxl.1,id=root_port13,chassis=0,slot=2  \
-device cxl-type3,bus=root_port13,persistent-memdev=pmem0,lsa=cxl-lsa0,id=cxl-pmem0 \
-M cxl-fmw.0.targets.0=cxl.1,cxl-fmw.0.size=4G
```
```
# uname -r
6.3.12-200.fc38.x86_64
```

```
# cxl list -v
{
    "bus":"root0",
    "provider":"ACPI.CXL",
    "nr_dports":1,
    "dports":[
      {
        "dport":"pci0000:0c",
        "alias":"ACPI0016:00",
        "id":12
      }
    ],
    "ports:root0":[
      {
        "port":"port1",
        "host":"pci0000:0c",
        "depth":1,
        "nr_dports":1,
        "dports":[
          {
            "dport":"0000:0c:00.0",
            "id":0
          }
        ],
        "endpoints:port1":[
          {
            "endpoint":"endpoint2",
            "host":"mem0",
            "parent_dport":"0000:0c:00.0",
            "depth":2,
            "memdev":{
              "memdev":"mem0",
              "pmem_size":268435456,
              "serial":0,
              "host":"0000:0d:00.0"
            }
          }
        ],
        "decoders:port1":[
          {
            "decoder":"decoder1.0",
            "resource":19595788288,
            "size":268435456,
            "interleave_ways":1,
            "nr_targets":1,
            "targets":[
              {
                "target":"0000:0c:00.0",
                "position":0,
                "id":0
              }
            ]
          }
        ]
      }
    ],
    "decoders:root0":[
      {
        "decoder":"decoder0.0",
        "resource":19595788288,
        "size":4294967296,
        "interleave_ways":1,
        "max_available_extent":4294967296,
        "pmem_capable":true,
        "volatile_capable":true,
        "accelmem_capable":true,
        "nr_targets":1,
        "targets":[
          {
            "target":"pci0000:0c",
            "alias":"ACPI0016:00",
            "position":0,
            "id":12
          }
        ]
      }
    ]
  }


```

```
# lspci -vvv -s 
0d:00.0 CXL: Intel Corporation Device 0d93 (rev 01) (prog-if 10 [CXL Memory Device (CXL 2.x)])
	Subsystem: Red Hat, Inc. Device 1100
	Physical Slot: 2
	Control: I/O+ Mem+ BusMaster+ SpecCycle- MemWINV- VGASnoop- ParErr- Stepping- SERR+ FastB2B- DisINTx+
	Status: Cap+ 66MHz- UDF- FastB2B- ParErr- DEVSEL=fast >TAbort- <TAbort- <MAbort- >SERR- <PERR- INTx-
	Latency: 0
	Region 0: Memory at fe800000 (64-bit, non-prefetchable) [size=64K]
	Region 2: Memory at fe810000 (64-bit, non-prefetchable) [size=4K]
	Region 4: Memory at fe811000 (32-bit, non-prefetchable) [size=4K]
	Capabilities: [40] MSI-X: Enable+ Count=6 Masked-
		Vector table: BAR=4 offset=00000000
		PBA: BAR=4 offset=00000800
	Capabilities: [80] Express (v2) Endpoint, MSI 00
		DevCap:	MaxPayload 128 bytes, PhantFunc 0, Latency L0s <64ns, L1 <1us
			ExtTag- AttnBtn- AttnInd- PwrInd- RBE+ FLReset- SlotPowerLimit 0W
		DevCtl:	CorrErr+ NonFatalErr+ FatalErr+ UnsupReq+
			RlxdOrd- ExtTag- PhantFunc- AuxPwr- NoSnoop-
			MaxPayload 128 bytes, MaxReadReq 128 bytes
		DevSta:	CorrErr- NonFatalErr- FatalErr- UnsupReq- AuxPwr- TransPend-
		LnkCap:	Port #0, Speed 2.5GT/s, Width x1, ASPM L0s, Exit Latency L0s <64ns
			ClockPM- Surprise- LLActRep- BwNot- ASPMOptComp-
		LnkCtl:	ASPM Disabled; RCB 64 bytes, Disabled- CommClk-
			ExtSynch- ClockPM- AutWidDis- BWInt- AutBWInt-
		LnkSta:	Speed 2.5GT/s, Width x1
			TrErr- Train- SlotClk- DLActive- BWMgmt- ABWMgmt-
		DevCap2: Completion Timeout: Not Supported, TimeoutDis- NROPrPrP- LTR-
			 10BitTagComp- 10BitTagReq- OBFF Not Supported, ExtFmt+ EETLPPrefix+, MaxEETLPPrefixes 4
			 EmergencyPowerReduction Not Supported, EmergencyPowerReductionInit-
			 FRS- TPHComp- ExtTPHComp-
			 AtomicOpsCap: 32bit- 64bit- 128bitCAS-
		DevCtl2: Completion Timeout: 50us to 50ms, TimeoutDis- LTR- 10BitTagReq- OBFF Disabled,
			 AtomicOpsCtl: ReqEn-
		LnkCtl2: Target Link Speed: 2.5GT/s, EnterCompliance- SpeedDis-
			 Transmit Margin: Normal Operating Range, EnterModifiedCompliance- ComplianceSOS-
			 Compliance Preset/De-emphasis: -6dB de-emphasis, 0dB preshoot
		LnkSta2: Current De-emphasis Level: -6dB, EqualizationComplete- EqualizationPhase1-
			 EqualizationPhase2- EqualizationPhase3- LinkEqualizationRequest-
			 Retimer- 2Retimers- CrosslinkRes: unsupported
	Capabilities: [100 v1] Designated Vendor-Specific: Vendor=1e98 ID=0000 Rev=1 Len=56: CXL
		CXLCap:	Cache- IO+ Mem+ Mem HW Init+ HDMCount 1 Viral-
		CXLCtl:	Cache- IO+ Mem+ Cache SF Cov 0 Cache SF Gran 0 Cache Clean- Viral-
		CXLSta:	Viral-
		CXLSta2:	ResetComplete+ ResetError- PMComplete-
		Cache Size Not Reported
		Range1: 0000000000000000-000000000fffffff
			Valid+ Active+ Type=CDAT Class=CDAT interleave=0 timeout=1s
		Range2: 0000000000000000-ffffffffffffffff
			Valid- Active- Type=Volatile Class=DRAM interleave=0 timeout=1s
	Capabilities: [138 v1] Designated Vendor-Specific: Vendor=1e98 ID=0008 Rev=0 Len=36: CXL
		Block1: BIR: bar0, ID: component registers, offset: 0000000000000000
		Block2: BIR: bar2, ID: CXL device registers, offset: 0000000000000000
	Capabilities: [15c v1] Designated Vendor-Specific: Vendor=1e98 ID=0005 Rev=0 Len=16: CXL
		GPF Phase 2 Duration: 3s
		GPF Phase 2 Power: 51mW
	Capabilities: [16c v1] Designated Vendor-Specific: Vendor=1e98 ID=0007 Rev=1 Len=20: CXL
		FBCap:	Cache- IO+ Mem+ 68BFlit+ MltLogDev-
		FBCtl:	Cache- IO+ Mem- SynHdrByp- DrftBuf- 68BFlit- MltLogDev- RCD- Retimer1- Retimer2-
		FBSta:	Cache- IO+ Mem+ SynHdrByp- DrftBuf- 68BFlit+ MltLogDev-
		FBModTS:	Received FB Data: 0000ef
	Capabilities: [190 v1] Data Object Exchange
		DOECap: IntSup+
			Interrupt Message Number 000
		DOECtl: IntEn-
		DOESta: Busy- IntSta- Error- ObjectReady-
	Capabilities: [200 v2] Advanced Error Reporting
		UESta:	DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
		UEMsk:	DLP- SDES- TLP- FCP- CmpltTO- CmpltAbrt- UnxCmplt- RxOF- MalfTLP- ECRC- UnsupReq- ACSViol-
		UESvrt:	DLP+ SDES+ TLP- FCP+ CmpltTO- CmpltAbrt- UnxCmplt- RxOF+ MalfTLP+ ECRC- UnsupReq- ACSViol-
		CESta:	RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr-
		CEMsk:	RxErr- BadTLP- BadDLLP- Rollover- Timeout- AdvNonFatalErr+
		AERCap:	First Error Pointer: 00, ECRCGenCap+ ECRCGenEn- ECRCChkCap+ ECRCChkEn-
			MultHdrRecCap- MultHdrRecEn- TLPPfxPres- HdrLogCap-
		HeaderLog: 00000000 00000000 00000000 00000000
	Kernel driver in use: cxl_pci
	Kernel modules: cxl_pci
```

**Comment by shadowcollecter** [comment]
adding the debug flag , it shows
```
[root@localhost ~]# cxl create-region -d decoder0.0 -t pmem -m mem0 --debug
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices' dev: 'mem'
libcxl: add_cxl_memdev: mem0: base: '/sys/bus/cxl/devices/mem0'
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/mem0' dev: 'pmem'
libcxl: add_cxl_pmem: pmem0: pmem_base: '/sys/bus/cxl/devices/mem0/pmem0'
libcxl: __sysfs_device_parse: pmem0: processed
libcxl: __sysfs_device_parse: mem0: processed
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices' dev: 'root'
libcxl: add_cxl_bus: root0: base: '/sys/bus/cxl/devices/root0'
libcxl: __util_modalias_to_module: failed to find module for alias: cxl:t4 0 list: empty
libcxl: __sysfs_device_parse: root0: processed
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0' dev: 'endpoint'
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0' dev: 'port'
libcxl: add_cxl_port: port1: base: '/sys/bus/cxl/devices/root0/port1'
libcxl: __util_modalias_to_module: alias: cxl:t3 module: cxl_port
libcxl: __sysfs_device_parse: port1: processed
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0/port1' dev: 'endpoint'
libcxl: add_cxl_endpoint: endpoint2: base: '/sys/bus/cxl/devices/root0/port1/endpoint2'
libcxl: __util_modalias_to_module: alias: cxl:t3 module: cxl_port
libcxl: __sysfs_device_parse: endpoint2: processed
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0' dev: 'decoder0.'
libcxl: add_cxl_decoder: decoder0.0: base: '/sys/bus/cxl/devices/root0/decoder0.0'
libcxl: __sysfs_read_attr: failed to open /sys/bus/cxl/devices/root0/decoder0.0/mode: No such file or directory
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0/decoder0.0' dev: 'region'
libcxl: add_cxl_decoder: decoder0.0: target0 /sys/devices/pci0000:0c phys_path: none
libcxl: add_cxl_decoder: decoder0.0: target0 /sys/devices/pci0000:0c fw_path: /sys/devices/LNXSYSTM:00/LNXSYBUS:00/ACPI0016:00
libcxl: __sysfs_device_parse: decoder0.0: processed
libcxl: cxl_target_maps_memdev: memdev: /sys/devices/pci0000:0c/0000:0c:00.0/0000:0d:00.0 target: /sys/devices/pci0000:0c
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0/port1' dev: 'decoder1.'
libcxl: add_cxl_decoder: decoder1.0: base: '/sys/bus/cxl/devices/root0/port1/decoder1.0'
libcxl: __sysfs_read_attr: failed to open /sys/bus/cxl/devices/root0/port1/decoder1.0/mode: No such file or directory
libcxl: add_cxl_decoder: decoder1.0: target0 /sys/devices/pci0000:0c/0000:0c:00.0 phys_path: none
libcxl: add_cxl_decoder: decoder1.0: target0 /sys/devices/pci0000:0c/0000:0c:00.0 fw_path: none
libcxl: __sysfs_device_parse: decoder1.0: processed
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0/port1' dev: 'port'
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0/decoder0.0' dev: 'region'
libcxl: add_cxl_region: region0: base: '/sys/bus/cxl/devices/root0/decoder0.0/region0'
libcxl: __util_modalias_to_module: alias: cxl:t6 module: cxl_core
libcxl: __sysfs_device_parse: region0: processed
libcxl: cxl_wait_probe: waited 71 milliseconds...
libcxl: __sysfs_device_parse: base: '/sys/bus/cxl/devices/root0/port1/endpoint2' dev: 'decode[ 1624.114287] cxl_region region0: Failed to synchronize CPU cache state
r2.'
libcxl: add_cxl_decoder: decoder2.0: base: '/sys/bus/cxl/devices/root0/port1/endpoint2/decoder2.0'
libcxl: __sysfs_read_attr: failed to open /sys/bus/cxl/devices/root0/port1/endpoint2/decoder2.0/target_list: No such file or directory
libcxl: __sysfs_device_parse: decoder2.0: processed
libcxl: write_attr: failed to open /sys/bus/cxl/drivers/cxl_nvdimm/bind: No such file or directory
libcxl: write_attr: failed to open /sys/bus/cxl/drivers/cxl_pmem_region/bind: No such file or directory
libcxl: write_attr: failed to open /sys/bus/cxl/drivers/cxl_nvdimm_bridge/bind: No such file or directory
libcxl: __util_bind: region0: bind failed
libcxl: cxl_region_enable: region0: failed to enable
cxl region: create_region: region0: failed to enable: No such device or address
cxl region: cmd_create_region: created 0 regions
```

**Comment by djbw** [comment]
tl;dr: Rebuild your kernel with CXL_REGION_INVALIDATION_TEST=y

Longer story:
CXL manages "host-physical" (HPA) address space. When HPA is reconfigured, CPU caches need to be flushed because they may have data cached from the memory that was previously at that HPA. On x86 the method to manage caches is the wbinvd instruction. The wbinvd instruction is invalid in QEMU, and the CXL driver in QEMU is only managing emulated host-physical-addresses. The CXL_REGION_INVALIDATION_TEST=y option tells the CXL core to ignore the fact that it is not running in a real CXL environment.

As for the hang with v6.5-rc that sounds like a bug. Can you check if the kernel crashed and post that backtrace? You may need to enable the serial port in QEMU and route the kernel messages there to capture the crash log.

**Comment by shadowcollecter** [comment]
the message is 
`
[ 1513.016337] cxl region0: Failed to synchronize CPU cache state
[ 1674.002413] cxl_region: probe of region1 failed with error -4
`
the 'dmesg ' didn't show  other stuff, after a period of time, I use control + c to stop it

I am new to the Linux kernel field, is the message  correct?

**Comment by shadowcollecter** [comment]
After change the kernel config , it works !
